Add cloud datasources

davidcassany opened the pull request:
This commit adds datasources from several providers (mostly cloud
providers). The datasource yip plugin is based on Linuxkit, however
the metadata utility from Linuxkit had to be forked and slightly
refactored to make it usable as a golang package.

Related to rancher-sandbox/cOS-toolkit#93

Signed-off-by: David Cassany <dcassany@suse.com>

davidcassany commented:
@mudler this is my take on how it would use linuxkit/pkg/metadata utility so it can be used as a golang package directly without touching anything (from yip POV). For that I had to patch, in fact I just had to rearrange some code structure (no new logic) and for that start moving I did a fork linuxkit.

Honestly I am surprised by the linuxkit project structure in general and by the `pkg/metadata` utility structure too. Looks like this is providing a toolset that should be reusable however the code is does not seam to be organized for that purpose... :man_shrugging:  Anyway, I thought that the fork approach is simpler that vendoring linuxkit and locally patch it. So up to that point, if we want to continue with this sort of implementation I see two paths, try to push my fork upstream (and use the fork meanwhile) and/or directly drop linuxkit and implement our own provider wrappers.

Another option I did not test but that should easy to implement is to actually make yip call the linuxkit metadata binary, so basically add a runtime binary dependency on yip to the metadata binary and just execute it from yip. To my eyes is the only possible option to reuse linuxkit without having to modify it.

IMHO, unless we have some good interaction with linuxkit and some changes on their side are applied I'd say we'd better implement our own code for several data source providers.

mudler commented:
@davidcassany totally agree, for the moment being it's fine and easier to maintain a fork indeed.

we can try to see with upstream if they are willing to take our patches. Since that code part does not seems to be actively developed anymore, the fork might be also an option that probably can just stick around for us :+1:

davidcassany commented:
@mudler could have a second round to review it?

You can read the two commits independently. The second one is what is new from what you already reviewed.

Note that I changed `files` and `dir` plugins. I added tests for both to check the new features.

mudler commented on pkg/plugins/dir.go:
```diff
@@ -30,3 +32,34 @@ func writeDirectory(dir schema.Directory, fs vfs.FS) error {
 	return fs.Chown(dir.Path, dir.Owner, dir.Group)
 }
+
+func writePath(dir schema.Directory, fs vfs.FS, topLevel bool) error {
```
I'm a bit confused here. The goal is to create all the path, like `mkdir -p`? In this case you could replace this call with [os.MkdirAll](https://golang.org/pkg/os/#MkdirAll)

davidcassany replied:
Well I did this way mostly to keep using the `fs vfs.FS` and have it all file and directory writes through the same interface.  I also I assumed to apply the provided permissions to any created directory, which is a design choice. I thought it is more natural to apply the provided ownership and permissions to any created directory rather than let it go with some default ownership. That's why I did not dig that much into the idea of using `MkdirAll` and thought about simply iterating over the already existing method.

mudler commented on pkg/plugins/files.go:
```diff
@@ -23,8 +25,32 @@ func EnsureFiles(s schema.Stage, fs vfs.FS, console Console) error {
+	_, err := fs.Stat(parentDir)
+	if err != nil {
+		log.Debug("Creating parent directories")
+		fmt.Printf("\nFilepermissions: %d\n", file.Permissions)
```
Leftover?

davidcassany replied:
Oh yes, my bad :facepalm:

mudler reviewed (approved):
Looks good here :+1:
